pkg/controller/imagelistpulljob/imagelistpulljob_controller.go

```diff
@@ -0,0 +1,422 @@
+			if unsatisfiedDuration >= expectations.ExpectationTimeout {
+				klog.Warningf("Expectation unsatisfied overtime for %v, timeout=%v", request.String(), unsatisfiedDuration)
+				return reconcile.Result{}, nil
+			}
+			klog.V(4).Infof("Not satisfied resourceVersion for %v", request.String())
+			return reconcile.Result{RequeueAfter: expectations.ExpectationTimeout - unsatisfiedDuration}, nil
+		}
+	}
+
+	// 3. Calculate the new status for this job
+	newStatus := r.calculateStatus(job, imagePullJobsMap)
+
+	// 4. Compute ImagePullJobActions
+	needToCreate, needToDelete := r.computeImagePullJobActions(job, imagePullJobsMap)
+
+	// 5. Sync ImagePullJob
+	err = r.syncImagePullJob(job, needToCreate, needToDelete)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// 6. Update status
+	if !util.IsJSONObjectEqual(&job.Status, newStatus) {
+		if err = r.updateStatus(job, newStatus); err != nil {
+			return reconcile.Result{}, fmt.Errorf("update ImageListPullJob status error: %v", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileImageListPullJob) updateStatus(job *appsv1alpha1.ImageListPullJob, newStatus *appsv1alpha1.ImageListPullJobStatus) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		imageListPullJob := &appsv1alpha1.ImageListPullJob{}
+		if err := r.Get(context.TODO(), types.NamespacedName{Namespace: job.Namespace, Name: job.Name}, imageListPullJob); err != nil {
+			return err
+		}
+		imageListPullJob.Status = *newStatus
+		return r.Status().Update(context.TODO(), imageListPullJob)
+	})
+}
+
+func (r *ReconcileImageListPullJob) computeImagePullJobActions(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) ([]*appsv1alpha1.ImagePullJob, []*appsv1alpha1.ImagePullJob) {
+	var needToDelete, needToCreate []*appsv1alpha1.ImagePullJob
+	//1. need to create
+	images, needToDelete := r.filterImagesAndImagePullJobs(job, imagePullJobs)
+	needToCreate = r.newImagePullJobs(job, images)
+	// some images delete from ImageListPullJob.Spec.Images
+	for image, imagePullJob := range imagePullJobs {
+		if !slice.ContainsString(job.Spec.Images, image, nil) {
+			needToDelete = append(needToDelete, imagePullJob)
+		}
+	}
+	return needToCreate, needToDelete
+}
+
+func (r *ReconcileImageListPullJob) calculateStatus(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) *appsv1alpha1.ImageListPullJobStatus {
+	var active, completed, succeeded int32
+	// record the failed image status
+	var failedImageStatuses []*appsv1alpha1.FailedImageStatus
+
+	for _, imagePullJob := range imagePullJobs {
+		if imagePullJob.Status.StartTime == nil {
+			continue
+		}
+
+		if imagePullJob.Status.Active > 0 {
+			active = active + 1
+		}
+
+		if imagePullJob.Status.Desired == (imagePullJob.Status.Failed + imagePullJob.Status.Succeeded) {
+			completed = completed + 1
+			//
+			if imagePullJob.Status.Failed > 0 {
```

Comment:
I think we can move 'if imagePullJob.Status.Failed > 0 {' to line252, then reduce cyclomatic complexity